hash_practice/exercises.py

```diff
@@ -22,8 +49,29 @@ def valid_sudoku(table):
         Each element can either be a ".", or a digit 1-9
         The same digit cannot appear twice or more in the same 
         row, column or 3x3 subgrid
-        Time Complexity: ?
-        Space Complexity: ?
+        Time Complexity: O(n**2)
+        Space Complexity: O(1)
     """
-    pass
-
+    # Note from Lizet to instructor: I know this doesn't pass the last 2 tests, but I'm not sure why.
+    # test cases but submitting as is since this last function is optional
+    # iterate through each row
+    for row in table:
+        # iterate through each element in the row
+        for element in row:
+            # if the element is not a "."
+            if element != ".":
+                # if the element is not a digit 1-9
+                if element not in "123456789":
+                    return False
+                # if the element is a digit 1-9
+                else:
+                    # if the element is in the same row
+                    if row.count(element) > 1:
+                        return False
+                    # if the element is in the same column
+                    if [row[i] for i in range(9)].count(element) > 1:
+                        return False
+                    # if the element is in the same 3x3 subgrid
+                    if [table[i][j] for i in range(3) for j in range(3)].count(element) > 1:

```

Comment:
And in this case, since `i` and `j` run through `range(3)` on every iteration, you're only going to get the top subgrid every time.

Rather than trying to be clever and keep the checks in the same nested loop as the row and column checks, I would recommend just having a separate loop for the subgrid checking. There are some ways you can use the `//` integral division operator to get the right subgrid, but since allow you to assume that all grids are 3x3, it may just be easiest to have a constant list of indices to work through (which is what is done in the instructor solution).

(You could also separate out the column checks from the row checks if you wanted to, but that is less necessary.)